### PR TITLE
Make sdl2-compat even more compatible with SDL2

### PR DESCRIPTION
**Pull request by madebr**
At this speed, sdl2-compat is going to be more compatible to sdl2 that sdl2 itself :P

- Fixes SOVERSION + DYLIB compatibility/current ersioning
- Fixes libraries in sdl2_compat.pc
- Fixes test/testgles2_sdf.c
- Fixes error trying to include X11 when it is not available

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -127,7 +127,9 @@ if(APPLE)
 #  set(DYLIB_COMPAT_VERSION 12.0.0 CACHE STRING "library compatibility version")
 #  set(DYLIB_CURRENT_VERSION 12.61.0 CACHE STRING "library current version")
   find_package(X11)
-  include_directories("${X11_INCLUDE_DIR}")
+  if(X11_FOUND)
+    include_directories("${X11_INCLUDE_DIR}")
```
Comment (sezero) [inline CMakeLists.txt:131]: Do you hit `X11_FOUND` with `X11_INCLUDE_DIR` set to something non-empty?

Comment (madebr) [reply CMakeLists.txt:131]: You should not: https://github.com/Kitware/CMake/blob/f72c405d4ee4ad4775a10a944774b91f49726dad/Modules/FindX11.cmake#L469-L471

File: CMakeLists.txt [hunk 2]
```diff
@@ -161,34 +194,32 @@ if(APPLE)
...
-    SOVERSION "0"
     OUTPUT_NAME "SDL2"
     PREFIX ""
   )
 elseif(OS2)
   set_target_properties(SDL2 PROPERTIES
-    VERSION "${PROJECT_VERSION}"
-    SOVERSION "0"
     OUTPUT_NAME "SDL2"
+    VERSION "${SDL_VERSION}"
```
Comment (sezero) [inline CMakeLists.txt:223]: _Should_ not be needed. (Also, since os/2 support is gone in sdl3, this part can go too.)

File: CMakeLists.txt (hunk 1 above)
Comment (sezero) [reply CMakeLists.txt:131]: Yes, and that's what I'm saying: `X11_INCLUDE_DIR` is wrong here. It should be `X11_X11_INCLUDE_PATH`, hoping that name it won't change across cmake versions.

Comment (sezero) [reply CMakeLists.txt:131]: To be more clear: You are _not_ using `FindX11` module which should set `X11_INCLUDE_DIR`. Instead, you are using `find_package()`.

Comment (madebr) [reply CMakeLists.txt:131]: But `X11_INCLUDE_DIR` is documented at [`FindX11`](https://cmake.org/cmake/help/v3.0/module/FindX11.html)?

`find_package(X11)` looks for `FindX11.cmake` or `X11Config.cmake` files and includes these.

Comment (madebr) [reply CMakeLists.txt:131]: [The search procedure is documented here](https://cmake.org/cmake/help/latest/command/find_package.html#search-modes)

Comment (sezero) [reply CMakeLists.txt:131]: OK then.

**Comment by sezero** [comment]
Tested building

- Natively for i686-linux, success: 
```
$ ls -l *.so*
lrwxrwxrwx. 1 ozzie ozzie     16 Nov 30 03:03 libSDL2-2.0.so -> libSDL2-2.0.so.0
lrwxrwxrwx. 1 ozzie ozzie     23 Nov 30 03:03 libSDL2-2.0.so.0 -> libSDL2-2.0.so.0.9000.0
-rwxrwxr-x. 1 ozzie ozzie 399872 Nov 30 03:03 libSDL2-2.0.so.0.9000.0
```

- Cross- for windows x86 and x86_64 using mingw and mingw-w64: Success.

- Cross- for macos x86_64 using clang and cctools-port: Success, but dylib
  compatibility and current versions still not set. Will you do that in this PR?
```
$ ls -l *.dylib
-rwxrwxr-x. 1 ozzie ozzie 387576 Nov 30 02:55 libSDL2-2.0.0.dylib
lrwxrwxrwx. 1 ozzie ozzie     19 Nov 30 02:55 libSDL2-2.0.dylib -> libSDL2-2.0.0.dylib

$ file libSDL2-2.0.0.dylib
libSDL2-2.0.0.dylib: Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>

$ x86_64-apple-darwin9-otool -L libSDL2-2.0.0.dylib
libSDL2-2.0.0.dylib:
	/home/ozzie/6/sdl2-compat/b/libSDL2-2.0.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 1265.21.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 48.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 855.17.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1056.16.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```
   Test programs unfathomably fail linkage, can't understand now, maybe later:
```
ld: warning: ignoring file libSDL2main.a, file was built for archive which is not the architecture being linked (x86_64): libSDL2main.a
ld: warning: ignoring file libSDL2_test.a, file was built for archive which is not the architecture being linked (x86_64): libSDL2_test.a
```

**Comment by madebr** [comment]
Does this fix the versioning?
```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -195,8 +195,8 @@ if(APPLE)
   # !!! FIXME: check OUTPUT_NAME value.
   set_target_properties(SDL2 PROPERTIES COMPILE_DEFINITIONS "_THREAD_SAFE")
   target_link_libraries(SDL2 PRIVATE
-    "-Wl,-compatibility_version,${SDL_DYLIB_COMPAT_VERSION_MICRO}"
-    "-Wl,-current_version,${SDL_DYLIB_CURRENT_VERSION_MICRO}"
+    "-Wl,-compatibility_version,${SDL_DYLIB_COMPAT_VERSION}"
+    "-Wl,-current_version,${SDL_DYLIB_CURRENT_VERSION}"
     "-Wl,-framework,AppKit"
   )
   set_target_properties(SDL2 PROPERTIES

```

**Comment by sezero** [comment]
Yes:
```
$ x86_64-apple-darwin9-otool -L libSDL2-2.0.0.dylib
libSDL2-2.0.0.dylib:
	/home/ozzie/6/sdl2-compat/b/libSDL2-2.0.0.dylib (compatibility version 9001.0.0, current version 9001.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 1265.21.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 48.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 855.17.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1056.16.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

The commented out lines around line 123 can go too.

**Comment by sezero** [comment]
> Test programs unfathomably fail linkage, can't understand now, maybe later:

Was my fault. Setting CMAKE_AR and CMAKE_RANLIB in the toolchain file fixed everything.

**Comment by madebr** [comment]
I pushed to main as 06f3adb

**Comment by madebr** [comment]
There is still a potential problem with the `.m` files because we don't do anything with the `OBJC` compiler or adding `-x objective-c`.

**Comment by sezero** [comment]
> There is still a potential problem with the `.m` files because we don't do anything with the `OBJC` compiler or adding `-x objective-c`.

That part is beyond my cmake skills (or lack thereof.)

**Comment by icculus** [comment]
Fwiw, this is what `cmake -G Ninja` produced to compile sdl2_compat_objc.m:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DSDL2_EXPORTS -D_THREAD_SAFE -I/usr/X11R6/include -I/usr/local/include -Wall -fvisibility=hidden -Wdeclaration-after-statement -Werror=declaration-after-statement -DSDL_DISABLE_IMMINTRIN_H -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk -mmacosx-version-min=12.6 -fPIC -MD -MT CMakeFiles/SDL2.dir/src/sdl2_compat_objc.m.o -MF CMakeFiles/SDL2.dir/src/sdl2_compat_objc.m.o.d -o CMakeFiles/SDL2.dir/src/sdl2_compat_objc.m.o -c /Users/icculus/projects/sdl2-compat/src/sdl2_compat_objc.m
```

This didn't force `-x objective-c` or whatever, but happens to work because GCC and Clang already know what to do with .m files.

And fwiw, testsprite2 worked with sdl2-compat on the Mac.  :)

**Comment by sezero** [comment]
> And fwiw, testsprite2 worked with sdl2-compat on the Mac. :)

Oh yes.

Out of curiosity: How much time does it take to compile sdl2_compat.c?
Haven't tested anywhere else yet, but on my broken motherboard + core2
duo + i686 linux with very old clang 3.4.2 against 10.9 sdk for x86_64,
it takes about a crazy 4m 35s. Possibly something wrong with my setup /
tools. (OTOH, sdl12_compat takes about 18s..)

**Comment by icculus** [comment]
I'd guess something in SDL_dynapi.c causes very old clang to hit a slow path...?

Or it's the huge amount of test apps?

**Comment by sezero** [comment]
No, only the library itself.

**Pull request merged**